File: rust/2019/day11.rs
use std::collections::HashMap;

mod util;
mod computer;

fn main() {
    let lines = util::get_lines();

    part1(&lines);
    part2(&lines);
}

fn part1(lines: &Vec<String>) {
    let program: Vec<i64> = lines[0].split(",").map(|n| n.parse::<i64>().unwrap()).collect();

    let mut c = computer::Computer::new(&program);
    let mut colors: HashMap<(i32, i32), i32> = HashMap::new();

    let dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)];
    let mut dir = 0;
    let mut pos = (0, 0);

    loop {
        let color: i32 = *colors.get(&pos).unwrap_or(&0);
        c.process(color as i64);

        if c.done() {
            break;
        }

        let paint = c.read() as i32;
        colors.insert(pos, paint);

        c.run_next();
        let turn = c.read() as i32;
        if turn == 0 {
            dir = (dir+3) % 4;
        } else {
            dir = (dir+1) % 4;
        }

        let move_dir = dirs[dir];
        pos = (pos.0+move_dir.0, pos.1+move_dir.1);
    }

    println!("Part 1 answer {}", colors.keys().len());
}

fn part2(lines: &Vec<String>) {
    let program: Vec<i64> = lines[0].split(",").map(|n| n.parse::<i64>().unwrap()).collect();

    let mut c = computer::Computer::new(&program);
    let mut colors: HashMap<(i32, i32), i32> = HashMap::new();

    let dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)];
    let mut dir = 0;
    let mut pos = (0, 0);

    colors.insert(pos, 1);

    loop {
        let color: i32 = *colors.get(&pos).unwrap_or(&0);
        c.process(color as i64);

        if c.done() {
            break;
        }

        let paint = c.read() as i32;
        colors.insert(pos, paint);

        c.run_next();
        let turn = c.read() as i32;
        if turn == 0 {
            dir = (dir+3) % 4;
        } else {
            dir = (dir+1) % 4;
        }

        let move_dir = dirs[dir];
        pos = (pos.0+move_dir.0, pos.1+move_dir.1);
    }

    for y in 0..10 {
        let mut line = String::new();
        for x in 0..50 {
            if *colors.get(&(y, x)).unwrap_or(&0) == 1 {
                line.push('#');
            } else {
                line.push(' ');
            }
        }
        println!("{}", line);
    }

    println!("Part 2 answer {}", colors.keys().len());
}

File: julia/2021/day11.jl
using Printf

include("util.jl")

# read relative to this dir
cd(@__DIR__)

input = [[parse(Int, c) for c in line] for line in readlines("input/day11.txt")]


## part 1

function flash(octo, x, y)
    flashes = 1

    octo[y][x] = 0
    for y2 in (y-1):(y+1)
        for x2 in (x-1):(x+1)
            if y2 < 1 || y2 > length(octo) || x2 < 1 || x2 > length(octo[x])
                continue
            end
            if octo[y2][x2] == 0
                continue
            end
            octo[y2][x2] += 1
        end
    end

    for y2 in (y-1):(y+1)
        for x2 in (x-1):(x+1)
            if y2 < 1 || y2 > length(octo) || x2 < 1 || x2 > length(octo[x])
                continue
            end
            if octo[y2][x2] > 9
                flashes += flash(octo, x2, y2)
            end
        end
    end
    
    flashes
end

function part1()
    octo = deepcopy(input)

    flashes = 0
    for iteration in 1:100
        for y in 1:length(octo)
            for x in 1:length(octo[y])
                octo[y][x] += 1
            end
        end

        for y in 1:length(octo)
            for x in 1:length(octo[y])
                if octo[y][x] > 9
                    flashes += flash(octo, x, y)
                end
            end
        end
    end

    println(flashes)
end
part1()

## part 2

function part2()    
    octo = deepcopy(input)

    for iteration in 1:10000
        flashes = 0
        for y in 1:length(octo)
            for x in 1:length(octo[y])
                octo[y][x] += 1
            end
        end

        for y in 1:length(octo)
            for x in 1:length(octo[y])
                if octo[y][x] > 9
                    flashes += flash(octo, x, y)
                end
            end
        end

        if flashes == 100
            print(iteration)
            break
        end
    end

    # 237 wrong
end
part2()